===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/compat.c,v
retrieving revision 1.39
retrieving revision 1.40
diff -u -r1.39 -r1.40
--- src/usr.bin/ssh/compat.c    2001/03/18 23:30:55     1.39
+++ src/usr.bin/ssh/compat.c    2001/03/23 11:04:06     1.40
@@ -23,7 +23,7 @@
  */

 #include "includes.h"
-RCSID("$OpenBSD: compat.c,v 1.39 2001/03/18 23:30:55 deraadt Exp $");
+RCSID("$OpenBSD: compat.c,v 1.40 2001/03/23 11:04:06 djm Exp $");

 #include <regex.h>

@@ -61,7 +61,9 @@
        } check[] = {
                { "^OpenSSH[-_]2\\.[012]",
                                        SSH_OLD_SESSIONID|SSH_BUG_BANNER },
-               { "^OpenSSH_2\\.3\\.0", SSH_BUG_BANNER },
+               { "^OpenSSH_2\\.3\\.0", SSH_BUG_BANNER|SSH_BUG_BIGENDIANAES },
+               { "^OpenSSH_2\\.5\\.[01]p1",
+                                       SSH_BUG_BIGENDIANAES },
                { "^OpenSSH",           0 },
                { "MindTerm",           0 },
                { "^2\\.1\\.0",         SSH_BUG_SIGBLOB|SSH_BUG_HMAC|
@@ -140,4 +142,34 @@
        }
        xfree(s);
        return ret;
+}
+
+char *
+compat_cipher_proposal(char *cipher_prop)
+{
+       char *orig_prop, *fix_ciphers;
+       char *cp, *tmp;
+       size_t len;
+
+       if (!(datafellows & SSH_BUG_BIGENDIANAES))
+               return(cipher_prop);
+
+       len = strlen(cipher_prop) + 1;
+       fix_ciphers = xmalloc(len);
+       *fix_ciphers = '\0';
+       tmp = orig_prop = xstrdup(cipher_prop);
+       while((cp = strsep(&tmp, ",")) != NULL) {
+               if (strncmp(cp, "aes", 3) && strncmp(cp, "rijndael", 8)) {
+                       if (*fix_ciphers)
+                               strlcat(fix_ciphers, ",", len);
+                       strlcat(fix_ciphers, cp, len);
+               }
+       }
+       xfree(orig_prop);
+       debug2("Original cipher proposal: %s", cipher_prop);
+       debug2("Compat cipher proposal: %s", fix_ciphers);
+       if (!*fix_ciphers)
+               fatal("No available ciphers found.");
+
+       return(fix_ciphers);
 }

===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/compat.h,v
retrieving revision 1.18
retrieving revision 1.19
diff -u -r1.18 -r1.19
--- src/usr.bin/ssh/compat.h    2001/03/18 23:30:55     1.18
+++ src/usr.bin/ssh/compat.h    2001/03/23 11:04:06     1.19
@@ -21,7 +21,7 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-/* RCSID("$OpenBSD: compat.h,v 1.18 2001/03/18 23:30:55 deraadt Exp $"); */
+/* RCSID("$OpenBSD: compat.h,v 1.19 2001/03/23 11:04:06 djm Exp $"); */

 #ifndef COMPAT_H
 #define COMPAT_H
@@ -43,11 +43,13 @@
 #define SSH_BUG_PKOK           0x0200
 #define SSH_BUG_PASSWORDPAD    0x0400
 #define SSH_BUG_SCANNER                0x0800
+#define SSH_BUG_BIGENDIANAES   0x1000

 void    enable_compat13(void);
 void    enable_compat20(void);
 void    compat_datafellows(const char *s);
 int    proto_spec(const char *spec);
+char   *compat_cipher_proposal(char *cipher_prop);
 extern int compat13;
 extern int compat20;
 extern int datafellows;

===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/sshconnect2.c,v
retrieving revision 1.54
retrieving revision 1.55
diff -u -r1.54 -r1.55
--- src/usr.bin/ssh/sshconnect2.c       2001/03/12 22:02:02     1.54
+++ src/usr.bin/ssh/sshconnect2.c       2001/03/23 11:04:07     1.55
@@ -23,7 +23,7 @@
  */

 #include "includes.h"
-RCSID("$OpenBSD: sshconnect2.c,v 1.54 2001/03/12 22:02:02 markus Exp $");
+RCSID("$OpenBSD: sshconnect2.c,v 1.55 2001/03/23 11:04:07 djm Exp $");

 #include <openssl/bn.h>
 #include <openssl/md5.h>
@@ -95,6 +95,9 @@
                myproposal[PROPOSAL_MAC_ALGS_CTOS] =
                myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
        }
+
+       myproposal[PROPOSAL_ENC_ALGS_STOC] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);

        /* buffers with raw kexinit messages */
        server_kexinit = xmalloc(sizeof(*server_kexinit));

===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/sshd.c,v
retrieving revision 1.176
retrieving revision 1.177
diff -u -r1.176 -r1.177
--- src/usr.bin/ssh/sshd.c      2001/03/22 20:22:55     1.176
+++ src/usr.bin/ssh/sshd.c      2001/03/23 11:04:07     1.177
@@ -1423,6 +1423,9 @@
                myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
        }
        myproposal[PROPOSAL_SERVER_HOST_KEY_ALGS] = list_hostkey_types();
+
+       myproposal[PROPOSAL_ENC_ALGS_STOC] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);

        server_kexinit = kex_init(myproposal);
        client_kexinit = xmalloc(sizeof(*client_kexinit));

===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/sshconnect2.c,v
retrieving revision 1.59
retrieving revision 1.60
diff -u -r1.59 -r1.60
--- src/usr.bin/ssh/sshconnect2.c       2001/03/29 14:24:59     1.59
+++ src/usr.bin/ssh/sshconnect2.c       2001/03/29 21:06:21     1.60
@@ -85,6 +85,10 @@
                myproposal[PROPOSAL_ENC_ALGS_CTOS] =
                myproposal[PROPOSAL_ENC_ALGS_STOC] = options.ciphers;
        }
+       myproposal[PROPOSAL_ENC_ALGS_CTOS] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_CTOS]);
+       myproposal[PROPOSAL_ENC_ALGS_STOC] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);
        if (options.compression) {
                myproposal[PROPOSAL_COMP_ALGS_CTOS] =
                myproposal[PROPOSAL_COMP_ALGS_STOC] = "zlib";
@@ -96,9 +100,6 @@
                myproposal[PROPOSAL_MAC_ALGS_CTOS] =
                myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
        }
-
-       myproposal[PROPOSAL_ENC_ALGS_STOC] =
-           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);

        /* buffers with raw kexinit messages */
        server_kexinit = xmalloc(sizeof(*server_kexinit));

===================================================================
RCS file: /usr/OpenBSD/cvs/src/usr.bin/ssh/sshd.c,v
retrieving revision 1.183
retrieving revision 1.184
diff -u -r1.183 -r1.184
--- src/usr.bin/ssh/sshd.c      2001/03/28 21:59:41     1.183
+++ src/usr.bin/ssh/sshd.c      2001/03/29 21:06:21     1.184
@@ -1393,14 +1393,16 @@
                myproposal[PROPOSAL_ENC_ALGS_CTOS] =
                myproposal[PROPOSAL_ENC_ALGS_STOC] = options.ciphers;
        }
+       myproposal[PROPOSAL_ENC_ALGS_CTOS] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_CTOS]);
+       myproposal[PROPOSAL_ENC_ALGS_STOC] =
+           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);
+
        if (options.macs != NULL) {
                myproposal[PROPOSAL_MAC_ALGS_CTOS] =
                myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
        }
        myproposal[PROPOSAL_SERVER_HOST_KEY_ALGS] = list_hostkey_types();
-
-       myproposal[PROPOSAL_ENC_ALGS_STOC] =
-           compat_cipher_proposal(myproposal[PROPOSAL_ENC_ALGS_STOC]);

        server_kexinit = kex_init(myproposal);
        client_kexinit = xmalloc(sizeof(*client_kexinit));

